Fb test sample create sm integration

labkey-chrisj opened the pull request:
#### Rationale
This adds support for SM integration to Biologics

#### Related Pull Requests
https://github.com/LabKey/biologics/pull/678

#### Changes
updates to EntityInsertPanel, EditableGrid

labkey-chrisj committed:
Add test coverage for SM integration, Sample Create

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_testSampleCreate_SMIntegration

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_testSampleCreate_SMIntegration

labkey-chrisj committed:
add ability to get parent select from test code

labkey-chrisj committed:
Implement doAndWaitForUpdate on EditableGrid

labkey-chrisj committed:
expose remove-rows button in component

labkey-tchad commented on src/org/labkey/test/components/glassLibrary/grids/EditableGrid.java:
```diff
@@ -512,6 +512,24 @@ public boolean isDisplayed()
         }
     }
 
+    @Override
+    protected void clearElementCache()
+    {
+        super.clearElementCache();
+    }
+
```
Why override `clearElementCache`? This doesn't change the functionality or access level.
```suggestion
```

labkey-chrisj replied:
I was following the pattern, looks like I got it wrong.  I'll take this change

labkey-chrisj replied:
Done

labkey-tchad commented on src/org/labkey/test/components/ui/EntityBulkInsertDialog.java:
```diff
@@ -88,6 +91,43 @@ public String getFieldWithId(String id)
+        return getCheckBox(fieldKey).get();
+    }
+
+    private Checkbox getCheckBox(String fieldKey)
+    {
+        WebElement row = elementCache().formRow(fieldKey);
+        return new Checkbox(elementCache().checkBoxLoc.findElement(row));
+    }
+
```
Add `setCheckbox`?

labkey-danield replied:
I understand the getCheckBox method but I think the name is confusing. Why not change it to getCheckboxElement?

labkey-chrisj replied:
It's already handled in `setBooleanField , getBooleanField`

labkey-tchad commented on src/org/labkey/test/components/ui/EntityBulkUpdateDialog.java:
```diff
@@ -104,6 +104,25 @@ public String getNumericField(String fieldKey)
         return elementCache().numericInput(fieldKey).get();
     }
 
+    public EntityBulkUpdateDialog setDateField(String fieldKey, String dateString)
+    {
+        Input input = elementCache().textInput(fieldKey);
+        elementCache().title.click();
+        getWrapper().waitFor(()-> !isDatePickerExpanded(), 1500);
+        input.set(dateString);
+        return this;
+    }
+
+    private boolean isDatePickerExpanded()
+    {
+        return Locator.tagWithClass("div", "react-datepicker-popper").existsIn(this);
+    }
```
Consider adding a `ReactDateInput` that handles waiting for the date picker popup to disappear after setting.

labkey-tchad replied:
Actually, is this a different field type than in `EntityBulkInsertDialog.setDateField`? They do different things so I'm wondering if one or the other is an earlier implementation or if they are actually different components.

labkey-chrisj replied:
Functionally, the datepicker here appears to be the same in both dialogs.

I've implemented a component class to wrap it- right now it exposes get/set/clear methods that seem to work in this context.
My initial (in-dialog) means of closing the datePicker popup was to click on the dialog title banner; that's a safe spot that won't dismiss the dialog.  Right now, the component isn't dialog-aware, so it clicks half the height of the edit box above or below, depending on whether the datepicker popup is below vs. above.

labkey-chrisj replied:
I'll assume for now you meant that I should create a ReactDateInput that extends input, and handles the popup

labkey-chrisj replied:
Unfortunately, it's hard to close the datepicker if you aren't either clicking in a safe spot outside of it, or selecting a time from the time-list.  I've got an attempt at a basic component that works so far in the use cases I have for it.

labkey-tchad commented on src/org/labkey/test/components/ui/EntityInsertPanel.java:
```diff
@@ -190,8 +204,13 @@ protected ElementCache newElementCache()
         WebElement bulkInsert = Locator.button("Bulk Insert").findWhenNeeded(_driver);
-        WebElement deleteRows = Locator.button("Delete rows").findWhenNeeded(_driver);
+        WebElement bulkUpdate = Locator.button("Bulk update").findWhenNeeded(_driver);
+        WebElement deleteRows = deleteRowsBtnLoc.findWhenNeeded(_driver);
```
Should the elements in this cache be scoped to the component instead of the entire page? e.g.
```suggestion
        WebElement deleteRows = deleteRowsBtnLoc.findWhenNeeded(this);
```

labkey-chrisj replied:
I've scoped these finders down to the insertPanel

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_testSampleCreate_SMIntegration

labkey-danield commented on src/org/labkey/test/components/ui/EntityBulkInsertDialog.java:
```diff
@@ -88,6 +91,43 @@ public String getFieldWithId(String id)
+        input.getComponentElement().sendKeys(Keys.ENTER);               // enter to collapse the datepicker
+        getWrapper().waitFor(()-> !isDatePickerExpanded(), 1500);
+        return this;
+    }
```
Should add a getDateField as well, other wise this is a "write only" field? Even if the getDateField calls getTextField (which it might) it makes the overall interface more consistent/coherent.

labkey-danield commented on src/org/labkey/test/components/ui/EntityBulkInsertDialog.java:
```diff
@@ -88,6 +91,43 @@ public String getFieldWithId(String id)
         return getWrapper().getFormElement(Locator.tagWithId("input", id));
     }
 
+    public EntityBulkInsertDialog setDateField(String fieldKey, String dateString)
```
We should be more consistent on parameter names for public functions. Examples of other public functions in this class are:
`public EntityBulkInsertDialog setTextField(String fieldCaption, String value)`
and
`public EntityBulkInsertDialog setFieldWithId(String id, String value)`
I'm not sure fieldCaption is a great option, it could be misleading. I find "fieldKey" a little confusing as well. I think we should use parameters name that indicates hints at what attribute the locator will use to find the field.

labkey-danield replied:
I'm still not a fan of fieldKey as a parameter name because it is not an attribute in the tag, id and name are.
But this is a discussion better had in the Test Team Working Session meeting.

labkey-chrisj replied:
FWIW, I like the term here because it is also literally the column name, in addition to being an attribute of the field row, and the ID of the inputs that aren't reactSelects.  For them, we're kinda stuck using label text.
Either way, I'm cool with calling the parameters whatever but yeah It's worth talking about

labkey-danield commented on src/org/labkey/test/components/ui/EntityBulkInsertDialog.java:
```diff
@@ -136,6 +176,19 @@ protected ElementCache elementCache()
+        {
+            WebElement inputEl = textInputLoc.findElement(elementCache().formRow(fieldKey));
+            return new Input(inputEl, getDriver());
+        }
```
Why not replace the locator in the getTextField and setTextField with this locator.

labkey-chrisj replied:
good call

labkey-chrisj replied:
TBH I'd avoided doing that because doing it would require me to crack open SM tests.
It turns out that getTextField as implemented took the column label text in the dialog, not the fieldKey/column name.
Buuuuut, I'll do it

labkey-danield replied:
Thanks!

labkey-chrisj committed:
create component wrapper for ReactDatePicker

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_testSampleCreate_SMIntegration

labkey-chrisj committed:
revert override of clearElementCache

labkey-chrisj committed:
CR feedback
- use datePicker for date field operations
- use fieldKey consistently for accessing text, number, and boolean fields
